source/common/ssl/context_manager_impl.cc

```diff
@@ -9,47 +9,55 @@
 namespace Envoy {
 namespace Ssl {
 
-ContextManagerImpl::~ContextManagerImpl() { ASSERT(contexts_.empty()); }
-
-void ContextManagerImpl::releaseContext(Context* context) {
-  std::unique_lock<std::shared_timed_mutex> lock(contexts_lock_);
+ContextManagerImpl::~ContextManagerImpl() {
+  removeEmptyContexts();
+  ASSERT(contexts_.empty());
+}
 
-  // context may not be found, in the case that a subclass of Context throws
-  // in it's constructor. In that case the context did not get added, but
-  // the destructor of Context will run and call releaseContext().
-  contexts_.remove(context);
+void ContextManagerImpl::removeEmptyContexts() {
+  contexts_.remove_if([](std::weak_ptr<Context> n) { return n.expired(); });
 }
 
-ClientContextPtr ContextManagerImpl::createSslClientContext(Stats::Scope& scope,
-                                                            const ClientContextConfig& config) {
-  ClientContextPtr context(new ClientContextImpl(*this, scope, config));
+ClientContextSharedPtr
+ContextManagerImpl::createSslClientContext(Stats::Scope& scope, const ClientContextConfig& config) {
+  ClientContextSharedPtr context = std::make_shared<ClientContextImpl>(scope, config);
   std::unique_lock<std::shared_timed_mutex> lock(contexts_lock_);
-  contexts_.emplace_back(context.get());
+  removeEmptyContexts();
+  contexts_.emplace_back(context);
   return context;
 }
 
-ServerContextPtr
+ServerContextSharedPtr
 ContextManagerImpl::createSslServerContext(Stats::Scope& scope, const ServerContextConfig& config,
                                            const std::vector<std::string>& server_names) {
-  ServerContextPtr context(new ServerContextImpl(*this, scope, config, server_names, runtime_));
+  ServerContextSharedPtr context =
+      std::make_shared<ServerContextImpl>(scope, config, server_names, runtime_);
   std::unique_lock<std::shared_timed_mutex> lock(contexts_lock_);
-  contexts_.emplace_back(context.get());
+  removeEmptyContexts();
+  contexts_.emplace_back(context);
   return context;
 }
 
 size_t ContextManagerImpl::daysUntilFirstCertExpires() const {
   std::shared_lock<std::shared_timed_mutex> lock(contexts_lock_);
   size_t ret = std::numeric_limits<int>::max();
-  for (Context* context : contexts_) {
-    ret = std::min<size_t>(context->daysUntilFirstCertExpires(), ret);
+  for (const auto& ctx_weak_ptr : contexts_) {
+    ContextSharedPtr context = ctx_weak_ptr.lock();
+    if (context) {
+      ret = std::min<size_t>(context->daysUntilFirstCertExpires(), ret);
+    }
   }
   return ret;
 }
 
 void ContextManagerImpl::iterateContexts(std::function<void(const Context&)> callback) {
   std::shared_lock<std::shared_timed_mutex> lock(contexts_lock_);
-  for (Context* context : contexts_) {
-    callback(*context);
+  removeEmptyContexts();
```

Comment:
Sorry I'm taking back my previous comment, let's not do this since we're removing in each create. `shared_lock` above is not enough to perform this.

Reply:
Done